### PR TITLE
[angular] Switch to Angular CLI

### LINKED ISSUE #10539: Move back to angular cli
##### **Overview of the feature request**

I would like to use angular cli in conjunction with jhipster,

##### **Motivation for or Use Case**

Angular CLI has well matured since it's start, and allows to easily enable features like Ivy...,
I suggest we switch back to using it.

##### **Related issues or PR**

https://github.com/jhipster/generator-jhipster/issues/9817 discussion started here, but as suggest by @wmarques I created a new Issue.

@deepu105  can this be addd to Hacktoberfest please ? Thanks

-   [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by wmarques**
Fixes #10539

As discussed, given the opportunities offered by the Angular CLI, we can now extend their config with ours so IMO we should move into Angular CLI.

This offers some advantages:
* Less config files
* Less dependencies
* Enjoy the benefits of Angular CLI (further optimisations, differential loading)

We could even go further with the Angular CLI schematics, for example we could do an Entity schematic to generate an entity page.

**Still Todo:**
* Compare bundles (check drawbacks)
* Remove old dependencies / useless files
* `ng test` Jest support
* PWA support

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by DanielFran** [comment]
@wmarques Can you fix conflicts?
Would be awesome to have it ready for next release.

**Comment by wmarques** [comment]
@DanielFran I think it will be for the v7, I want to test this well and to check if we don't lost performances

File: generators/client/templates/angular/package.json.ejs [hunk 1]
```diff
@@ -60,6 +60,8 @@
     "zone.js": "0.10.2"
   },
   "devDependencies": {
+    "@angular-builders/custom-webpack": "^8.2.0",
```
Comment (qmonmert) [inline generators/client/templates/angular/package.json.ejs]: "^8.2.0" => "8.2.0"

File: generators/client/templates/angular/package.json.ejs [hunk 2]
```diff
@@ -60,6 +60,8 @@
...
+    "@angular-devkit/build-angular": "^0.803.3",
```
Comment (qmonmert) [inline generators/client/templates/angular/package.json.ejs]: "^0.803.3" => "0.803.3"

File: generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs [hunk 3]
```diff
@@ -18,4 +18,4 @@
 -%>
 import 'core-js/proposals/reflect-metadata';
 import 'zone.js/dist/zone';
-require('../manifest.webapp');
+//require('../manifest.webapp');
```
Comment (qmonmert) [inline generators/client/templates/angular/src/main/webapp/app/polyfills.ts.ejs]: so you can remove this line ?

**Comment by mraible** [comment]
This PR needs a rebase to fix conflicts.

**Comment by wmarques** [comment]
Yep I'll work on it during the month

**Comment by pascalgrimaud** [comment]
@wmarques : what is the state of your PR ? The original ticket is opened for too long (8 months)

File: generators/client/templates/angular/tsconfig.app.json.ejs [hunk 4]
```diff
@@ -18,5 +18,5 @@
 -%>
 {
     "extends": "./tsconfig.json",
-    "files": ["<%= MAIN_SRC_DIR %>app/app.main.ts"]
+    "files": ["<%= MAIN_SRC_DIR %>app/app.main.ts","<%= MAIN_SRC_DIR %>app/polyfills.ts"]
```
Comment (qmonmert) [inline generators/client/templates/angular/tsconfig.app.json.ejs:21]: `"<%= MAIN_SRC_DIR %>app/app.main.ts","<%= MAIN_SRC_DIR %>app/polyfills.ts"` 
=>
`"<%= MAIN_SRC_DIR %>app/app.main.ts", "<%= MAIN_SRC_DIR %>app/polyfills.ts"`
😇

File: test/utils/expected-files.js [hunk 5]
```diff
@@ -375,10 +375,10 @@ const expectedFiles = {
 
     client: [
         'angular.json',
+        'ngsw-config.json',
         '.eslintrc.json',
         'package.json',
-        'postcss.config.js',
```
Comment (mshima) [inline test/utils/expected-files.js:380]: We need to delete every removed file in:
https://github.com/jhipster/generator-jhipster/blob/355fb16ea2d5feb08408083f1aaddfdb13bc461c/generators/cleanup.js#L107-L110

File: generators/client/templates/angular/tsconfig.spec.json.ejs [hunk 6]
```diff
@@ -16,8 +16,18 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 -%>
-module.exports = {
-    plugins: [
-        require('autoprefixer')
+{
+    "extends": "./tsconfig.json",
+    "compilerOptions": {
+        "outDir": "./out-tsc/spec",
```
Comment (mshima) [inline generators/client/templates/angular/tsconfig.spec.json.ejs:22]: This folder should be added to gitignore/eslintignore/prettierignore?

File: generators/client/templates/angular/webpack/proxy.conf.js.ejs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+                    '/management',
+                    '/swagger-resources',
+                    '/v2/api-docs',
+                    '/v3/api-docs',
+                    '/h2-console',
+                    <%_ if (authenticationType === 'oauth2') { _%>
+                    '/oauth2',
+                    '/login',
+                    <%_ } _%>
+                    '/auth'
```
Comment (mshima) [inline generators/client/templates/angular/webpack/proxy.conf.js.ejs:16]: Add `/health`?

File: generators/client/templates/angular/angular.json.ejs [hunk 8]
```diff
@@ -25,6 +25,80 @@
...
+          "builder": "@angular-builders/custom-webpack:browser",
+          "options": {
+            "customWebpackConfig": {
+              "path": "./webpack/webpack.custom.js"
+            },
+<%_ if (buildTool === 'gradle') { %>
+            "outputPath": "build/resources/main/static/",
+<%_ } else { %>
+            "outputPath": "target/classes/static",
+<%_ } %>
```
Comment (mshima) [inline generators/client/templates/angular/angular.json.ejs:39]:
```suggestion
            "outputPath": "<%= DIST_DIR %>",
```

File: generators/client/templates/angular/angular.json.ejs [hunk 9]
```diff
@@ -25,6 +25,80 @@
...
+            "index": "src/main/webapp/index.html",
+            "main": "src/main/webapp/app/app.main.ts",
+            "polyfills": "src/main/webapp/app/polyfills.ts",
```
Comment (mshima) [inline generators/client/templates/angular/angular.json.ejs:42]:
```suggestion
            "index": "<%= MAIN_SRC_DIR %>index.html",
            "main": "<%= MAIN_SRC_DIR %>app/app.main.ts",
            "polyfills": "<%= MAIN_SRC_DIR %>app/polyfills.ts",
```

File: generators/client/templates/angular/angular.json.ejs [hunk 10]
```diff
@@ -25,6 +25,80 @@
...
+            "aot": true,
+            "assets": [
+              "<%= MAIN_SRC_DIR %>content",
+              "<%= MAIN_SRC_DIR %>favicon.ico",
+              "<%= MAIN_SRC_DIR %>manifest.webapp",
+              { "glob": "*.{js,css,html,png}", "input": "./node_modules/swagger-ui-dist", "output": "swagger-ui", "ignore": ["**/index.html"]  },
+              { "glob": "axios.min.js", "input": "./node_modules/axios/dist", "output": "swagger-ui" },
+              { "glob": "**/*", "input": "<%= MAIN_SRC_DIR %>swagger-ui/", "output": "swagger-ui" }
+            ],
+            "styles": ["src/main/webapp/content/scss/global.scss", "src/main/webapp/content/scss/vendor.scss"],
```
Comment (mshima) [inline generators/client/templates/angular/angular.json.ejs:53]:
```suggestion
            "styles": ["<%= MAIN_SRC_DIR %>content/scss/global.scss", "<%= MAIN_SRC_DIR %>content/scss/vendor.scss"],
```

**Review by mshima — CHANGES_REQUESTED** [review]
Should `webpack.(common|prod|dev).js` be removed?

**Comment by wmarques** [comment]
Thanks alot @mshima for your review ! I'll do this tomorrow in the afternoon, hope we can merge this asap then

**Comment by mshima** [comment]
> Thanks alot @mshima for your review ! I'll do this tomorrow in the afternoon, hope we can merge this asap then

Saw yesterday that this is almost ready.
Great work.

**Comment by mshima** [comment]
@wmarques do you want some help finishing this PR?

**Comment by wmarques** [comment]
No thanks I'll apply your comments tomorrow, I was quite busy today 🙁

File: generators/client/templates/angular/tsconfig.spec.json.ejs (hunk 6 above)
Comment (wmarques) [reply generators/client/templates/angular/tsconfig.spec.json.ejs:22]: It's quite strange as I don't see any files written on the disk when I run unit tests...
so I don't know if we need to do this

File: test/utils/expected-files.js (hunk 5 above)
Comment (wmarques) [reply test/utils/expected-files.js:380]: Yes, I'll remove also the webpack files !

**Comment by wmarques** [comment]
I think I applied all your comments @mshima 
I also rebased, hope I didn't mess anything....

File: generators/client/files-angular.js [hunk 11]
```diff
@@ -29,17 +29,17 @@ const files = {
         {
             templates: [
                 'package.json',
-                'proxy.conf.json',
```
Comment (mshima) [inline generators/client/files-angular.js:32]: Remove in cleanup.js.

File: generators/client/files-angular.js [hunk 12]
```diff
@@ -360,8 +360,6 @@ const files = {
             path: CLIENT_TEST_SRC_DIR,
             templates: [
                 'jest.conf.js',
-                'jest.ts',
-                'jest-global-mocks.ts',
```
Comment (mshima) [inline generators/client/files-angular.js:364]: Remove in cleanup.js.

File: generators/client/templates/angular/tsconfig.spec.json.ejs (hunk 6 above)
Comment (mshima) [reply generators/client/templates/angular/tsconfig.spec.json.ejs:22]: Maybe change to
```suggestion
        "outDir": "<%= BUILD_DIR %>out-tsc/spec",
```

File: generators/client/templates/angular/webpack/proxy.conf.js.ejs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+                    '/v2/api-docs',
+                    '/v3/api-docs',
+                    '/h2-console',
+                    <%_ if (authenticationType === 'oauth2') { _%>
+                    '/oauth2',
+                    '/login',
+                    <%_ } _%>
+                    '/auth',
+                    '/health'
+                ],
```
Comment (mshima) [inline generators/client/templates/angular/webpack/proxy.conf.js.ejs:18]: Improve ident of this array.

**Review by mshima — CHANGES_REQUESTED** [review]
@wmarques there is a few more changes.
And we should check if recent dayjs migration is not broken.

**Comment by mshima** [comment]
Moment integration looks good.

**Comment by wmarques** [comment]
I'll apply the remaining comments later in few hours :)

**Comment by wmarques** [comment]
Should me good :) 
We should squash before merging I did alot of "fix" commits 😄

**Comment by mshima** [comment]
@wmarques there are 3 failures in `npm test` related to webpack.*.js removal.

File: generators/client/templates/angular/package.json.ejs [hunk 14]
```diff
@@ -93,41 +94,32 @@ limitations under the License.
     "chai-string": "1.5.0",
     <%_ } _%>
     "codelyzer": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "copy-webpack-plugin": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
```
Comment (kaidohallik) [inline generators/client/templates/angular/package.json.ejs:96]: `jhipster-needle-add-assets-to-webpack` is not currently implemented any more, restore this plugin and add this needle to `webpack.custom.js`? Or add this needle to `angular.json`?

File: generators/client/templates/angular/package.json.ejs [hunk 15]
```diff
@@ -155,14 +140,8 @@ limitations under the License.
...
     <%_ } _%>
-    "webpack": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
     "webpack-bundle-analyzer": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-cli": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-dev-server": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-merge": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-notifier": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "workbox-webpack-plugin": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "write-file-webpack-plugin": "VERSION_MANAGED_BY_CLIENT_ANGULAR"
+    "workbox-webpack-plugin": "VERSION_MANAGED_BY_CLIENT_ANGULAR"
```
Comment (kaidohallik) [inline generators/client/templates/angular/package.json.ejs:144]: This plugin is not used any more and can be removed.

File: generators/client/templates/angular/package.json.ejs [hunk 16]
```diff
@@ -155,14 +140,8 @@ limitations under the License.
     <%_ if (protractorTests) { _%>
     "webdriver-manager": "12.1.7",
     <%_ } _%>
-    "webpack": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
     "webpack-bundle-analyzer": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-cli": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-dev-server": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-merge": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
-    "webpack-notifier": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
```
Comment (kaidohallik) [inline generators/client/templates/angular/package.json.ejs:163]: Other frameworks are using `webpack-notifier`, shouldn't we keep using this also in Angular part? If dropping this functionality then needs to be deleted and cleaned up also file `webpack/logo-jhipster.png`.

File: generators/client/templates/angular/webpack/webpack.custom.js.ejs [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+module.exports = (config, options) => {
+  // RULES
+  if (config.mode === 'development') {
+    config.module.rules.push({
+      test: /\.(j|t)s$/,
+      enforce: 'pre',
+      loader: 'eslint-loader',
+      exclude: /node_modules/
+    });
+  }
```
Comment (kaidohallik) [inline generators/client/templates/angular/webpack/webpack.custom.js.ejs:21]: `eslint-loader` is deprecated and can be replaced with https://github.com/webpack-contrib/eslint-webpack-plugin (this can be also separate PR)

**Comment by wmarques** [comment]
Looks like it's a blueprint issue, let me check this

**Comment by wmarques** [comment]
jhipster-needle-add-assets-to-webpack
This needle is looking for `webpack.common.js` but it not longer exists.
I think I need to move this needle into `angular.json`

File: generators/cleanup.js [hunk 18]
```diff
@@ -120,10 +120,17 @@ function cleanupOldFiles(generator) {
             generator.removeFile(`${ANGULAR_DIR}shared/login/login.component.html`);
             generator.removeFile(`${ANGULAR_DIR}core/auth/user-route-access-service.ts`);
             generator.removeFile(`${ANGULAR_DIR}tsconfig.base.json`);
+            generator.removeFile(`${ANGULAR_DIR}webpack/webpack.common.js`);
+            generator.removeFile(`${ANGULAR_DIR}webpack/webpack.dev.js`);
+            generator.removeFile(`${ANGULAR_DIR}webpack/webpack.prod.js`);
```
Comment (kaidohallik) [inline generators/cleanup.js:125]: `webpack/utils.js` is no more needed and needs also to be deleted and cleaned up

**Review by kaidohallik — CHANGES_REQUESTED** [review]
Great work @wmarques !  
I have some comments about cleaning and keeping some functionality similar to other FE frameworks.

File: generators/client/templates/angular/package.json.ejs (hunk 14 above)
Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs:96]: Yes I'll check this, but I think the CLI also uses CopyWebpackPlugin so I don't know what the behaviour will be...

**Comment by DanielFran** [comment]
@wmarques Can you fix the conflicts.
There is too much changes in main repo, and this PR is almost done and should be merged as soon as you fix last comments 😊

File: generators/client/templates/angular/package.json.ejs (hunk 16 above)
Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs:163]: I don't see why I removed it, let's put it back. Btw it seems that we don't use it in Vue...

File: generators/client/templates/angular/webpack/webpack.custom.js.ejs (hunk 17 above)
Comment (wmarques) [reply generators/client/templates/angular/webpack/webpack.custom.js.ejs:21]: I think i'll do another PR for this :)

**Comment by wmarques** [comment]
@DanielFran Working on it !

File: generators/client/templates/angular/package.json.ejs (hunk 14 above)
Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs:96]: @kaidohallik I managed to have this working by including the CopyWebpackPlugin in webpack.custom.js, I think it's easier to manage than having to manipulate the angular.json file, WDYT ?

Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs:96]: I'm having this error because the copy array is invalid..
```
An unhandled exception occurred: Invalid options object. Copy Plugin has been initialized using an options object that does not match the API schema.
 - options.patterns should be an non-empty array.
```

Maybe I can put a dummy folder inside ?

**Comment by wmarques** [comment]
I'll fix this tomorrow, sorry I won't have time tonight 😢

File: generators/client/templates/angular/package.json.ejs (hunk 14 above)
Comment (kaidohallik) [reply generators/client/templates/angular/package.json.ejs:96]: Maybe this solution:
```
  const patterns = [
    // jhipster-needle-add-assets-to-webpack - JHipster will add/remove third-party resources in this array
  ];
  if (patterns.length > 0) {
    config.plugins.push(new CopyWebpackPlugin({ patterns }));
  }
```

Comment (wmarques) [reply generators/client/templates/angular/package.json.ejs:95]: That was so obvious... Thanks alot ! 😄

**Comment by pascalgrimaud** [comment]
@wmarques : if the CI is green and if you feel confident, go ahead and merge this. It will avoid some conflict and we can still do post review

File: test/needle-api/needle-client-webpack.spec.js [hunk 19]
```diff
@@ -50,8 +50,8 @@ const mockBlueprintSubGen = class extends ClientGenerator {
 };
 
 describe('needle API Webpack: JHipster client generator with blueprint', () => {
-    before(done => {
-        helpers
+    function generateAppWithClientFramework(clientFramework) {
+        return helpers
             .run(path.join(__dirname, '../../generators/client'))
```
Comment (mshima) [inline test/needle-api/needle-client-webpack.spec.js:55]:
```suggestion
            .create(path.join(__dirname, '../../generators/client'))
```

File: test/needle-api/needle-client-webpack.spec.js [hunk 20]
```diff
@@ -65,14 +65,23 @@ describe('needle API Webpack: JHipster client generator with blueprint', () => {
...
             .withPrompts({
                 baseName: 'jhipster',
-                clientFramework: ANGULAR,
+                clientFramework,
                 enableTranslation: true,
                 nativeLanguage: 'en',
                 languages: ['en', 'fr'],
-            })
-            .on('end', done);
+            });
```
Comment (mshima) [inline test/needle-api/needle-client-webpack.spec.js:72]:
```suggestion
            })
            .run();
```

**Review by mshima — CHANGES_REQUESTED** [review]

File: test/needle-api/needle-client-webpack.spec.js (hunk 19 above)
Comment (wmarques) [reply test/needle-api/needle-client-webpack.spec.js:55]: What's the difference between doing .run(..) or .create(..).run() ? :)

Comment (mshima) [reply test/needle-api/needle-client-webpack.spec.js]: `helpers.run`:
- uses root generator event listeners, so if others generators fails can create a false success or force the test to timeout.

`helpers.create`
- `run()` is a new api using Environment new promised apis.

Comment (wmarques) [reply test/needle-api/needle-client-webpack.spec.js]: Oh ok, didn't know about that, thanks for the explanation

**Comment by wmarques** [comment]
Finally !  Thanks guys, that was quite a journey 😄 
I'll keep an eye on issues and I'll provide some PRs to improve it (like the eslint-plugin as suggested by @kaidohallik)

**Comment by kaidohallik** [comment]
The final result LGTM, thanks for your hard work @wmarques !

**Comment by gmarziou** [comment]
I also think [the doc](https://www.jhipster.tech/using-angular/#-using-angular) and the generated README.md should be updated.

**Comment by wmarques** [comment]
@gmarziou you're right I'll work on it tonight
